2-Draft-Accepted/RFCxxxx-Secrets-Management.md

```diff
@@ -0,0 +1,150 @@
+New-PSSession -ComputerName myServer -Credential (Get-Secret -Name MyCreds)
+```
+
+## Specification
+
+### Secrets Vault
+
+There are two types of vaults: local and remote.
+A local vault, by default, is already created and named `Default`.
+`Default` on Windows is [CredMan](https://docs.microsoft.com/en-us/windows/desktop/SecAuthN/credentials-management).
+`Default` on macOS is [KeyChain](https://developer.apple.com/documentation/security/keychain_services).
+
+>[!NOTE]
+>KeyChain support is unlikely to be available in the first release of this feature.
+
+`Default` on Linux is [Hashicorp Vault](https://www.vaultproject.io/).
+
+>[!NOTE]
+>Linux has many options for local credential management, additional support may be
+>added in the future or via community provided extensions.
+
+### Credential Vault Extensions
+
+Vault extensions whether they are local or remote are provided as modules.
+
+Extensions are modules that provide a `Get-Secret` cmdlet that accepts a
+`-Credential` or `-AccessToken` parameter for authentication.
+Expectation is that each module will have their own cmdlet and not literally
+named `Get-Secret`.
+
+Extensions register themselves by calling `Register-SecretsVaultExtension`
+passing the name of the relevant cmdlet to `-Cmdlet`, the module name to `-Module`,
+and the module version to `-ModuleVersion`.
+This can be done as part of importing the module or explicitly by the user.
+
+`Register-SecretsVaultExtension` has a `-Default` switch to register the
+default local vault (if applicable, see above) if it has been unregistered.
+
+`Register-SecretsVaultExtension` takes a `-AccessToken` or `-Credential` parameter
+for authenticating with the remote vault.
+If neither are supplied, then the current user context is used and relies on
+the extension cmdlet to handle authentication.
+
+`Unregister-SecretsVaultExtension` should be used to remove any existing
+registration or when updating a registration to a new version.
+
+`Get-SecretsVaultExtension` will enumerate registered extensions returning
+the module name, module version, and cmdlet used.
+
+### Storing Secrets
+
+The `Register-Secret` cmdlet is used to store a secret.
+The `-Name` must be unique across all secrets regardless if local or remote.
+A `-Secret` parameter accepts an object that can be either a PSCredential
+or SecureString.
+
+### Retrieving Secrets
+
+The `Get-Secret` cmdlet is used to retrieve secrets.
+The `-Name` parameter is mandatory and retrieves the secret associated with
+that name.
+The returned object will either be a PSCredential or SecureString.
+
+### Authorization
+
+Access to the credential vault is always using the current process security context.
+In the case of remote vaults, the remote credential is stored within the local
+default vault and retrieved as needed when accessing secrets from that remote
+vault.
+
+### Secrets Supported
+
+Secrets supported will be:
+
+- PSCredential
+- SecureString
+
```

Comment:
It may also be important to consider end-to-end performance of the system.  If this can be used to retreive secure information quickly enough, it provides a good alternative to web.config when running PowerShell in IIS, and a good story where there is none when running PowerShell in a Net.HttpListener.  I believe a simple goal of being able to extract creds as plaintext from a local store within 2ms would be good.

Reply:
We'll have to do some perf measurements once we have code.  I don't know if 2ms is feasible as even plaintext would be encrypted.